### ISSUE TITLE
CentOS 7 Build fails to install zfs-initramfs rpm binary. 

### ISSUE THREAD

**Issue by tonymmm1** [openzfs/zfs#10885]
### System information

Type | Version/Name
 --- | ---
Distribution Name	|CentOS
Distribution Version	|7.8
Linux Kernel	|3.10.0-1127.13.1.el7.x86_64
Architecture	|x86_64
ZFS Version	|2.0.0-rc1_34_g6fffc88
SPL Version	|2.0.0-rc1_34_g6fffc88

### Describe the problem you're observing
Trying to compile dkms binaries like usual and noticed on CentOS7 it seems to fail to install the zfs-initramfs binary. 

```
--> Finished Dependency Resolution
Error: Package: zfs-initramfs-2.0.0-rc1_34_g6fffc88.el7.x86_64 (/zfs-initramfs-2.0.0-rc1_34_g6fffc88.el7.x86_64)
           Requires: initramfs-tools
```
### Describe how to reproduce the problem
Compiling dkms modules according to https://openzfs.github.io/openzfs-docs/Developer%20Resources/Custom%20Packages.html

Exact setup 

```
git clone https://github.com/openzfs/zfs.git
./autogen.sh
./configure.sh
make -j1 rpm-utils rpm-dkms
sudo yum localinstall *.$(uname -p).rpm *.noarch.rpm
```

**Comment by Ornias1993** [comment]
The docs have their own repo:
https://github.com/openzfs/openzfs-docs

It might be worth it closing this and making an issue there instead, just to make sure it isn't just a mistake in the documentation.

**Comment by tonymmm1** [comment]
Well the issue is not in the documentation it has something either to do with initramfs dependencies on ZFS and or CentOS packaging, etc.

**Comment by Ornias1993** [comment]
@tonymmm1 
You are very quick to come to that conclussion. What solutions did you try?

**Comment by tonymmm1** [comment]
```
[user1@s1-test ~]$ sudo yum search initramfs-tools
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.chpc.utah.edu
 * epel: ord.mirror.rackspace.com
 * extras: mirrors.xmission.com
 * updates: mirror.atl.genesisadaptive.com
Warning: No matches found for: initramfs-tools
No matches found
```
```
[user1@s1-test ~]$ rpm -qa | grep initramfs*
```

**Comment by tonymmm1** [comment]
I have compiled other version of zfs in the past with no issues so I thought it was strange that I was getting an error with the zfs-initramfs.rpm.

**Comment by Ornias1993** [comment]
> I have compiled other version of zfs in the past with no issues so I thought it was strange that I was getting an error with the zfs-initramfs.rpm.

Thats very well possible, but changes that invalidate the previous documentation can always come in allong the way. It doesn't mean there is a bug in code.

I'll take a look... the issue might be that a debian change has accidentally propegated into the rpm building side of things (initramfs-tools seems to be primarily a deb things as far as I can see quickly)

**Comment by tonymmm1** [comment]
I am trying to compile zfs-0.8.5-staging branch for now to see if there are any dependency issues with normal branch.

**Comment by Ornias1993** [comment]
Did you add all other packages as required by the general section of the doc article you linked?

**Comment by tonymmm1** [comment]
Yeah ran the command just now to double check.

**Comment by Ornias1993** [comment]
One thing:
make -j1 never worked right for me, please use the actual amount of threads but in my experience it does require at least 2 to run well.

**Comment by tonymmm1** [comment]
Really in my experience make usually works no matter what single threaded and usually breaks with higher job count.

**Comment by Ornias1993** [comment]
To be clear: that wasn't related to your question. But I had builds crash because of not having enough threads. (and yes, really, I checked it... I had to do 50 builds a day of ZFS at that time)

**Comment by Ornias1993** [comment]
Why are you suggesting you are building in accordance with the manual btw? That isn't true:

Manual:
```
$ cd zfs
$ ./configure
$ make -j1 rpm-utils rpm-dkms
$ sudo yum localinstall *.$(uname -p).rpm *.noarch.rpm
```

You:
```
git clone https://github.com/openzfs/zfs.git
./autogen.sh
./configure.sh
make -j1 rpm-utils rpm-dkms
sudo yum localinstall *.$(uname -p).rpm *.noarch.rpm
```

Firstoff:
You never actually went into the zfs folder after cloning the repo.
Second you used autogen, while the manual builds this without autogen.

**Comment by Ornias1993** [comment]
The full way of building/installing on CentOS 7 in accordance with the manual:

```
$ sudo yum install epel-release gcc make autoconf automake libtool rpm-build dkms libtirpc-devel libblkid-devel libuuid-devel libudev-devel openssl-devel zlib-devel libaio-devel libattr-devel elfutils-libelf-devel kernel-devel-$(uname -r) python python2-$ devel python-setuptools python-cffi libffi-devel
$ git clone https://github.com/openzfs/zfs.git
$ cd zfs
$ ./configure
$ make -j1 rpm-utils rpm-dkms
$ sudo yum localinstall *.$(uname -p).rpm *.noarch.rpm
```

**Comment by tonymmm1** [comment]
Yeah sorry it was implied lol.

**Comment by Ornias1993** [comment]
Okey...
If there is an issue, this PR would be the only one that could've caused it between the release of 0.8.4 and today:
https://github.com/openzfs/zfs/pull/10518

**Comment by tonymmm1** [comment]
Nope I did a full git clean and git reset as well as a pull for the 2.0.0 branch and have the same issue.
```
--> Finished Dependency Resolution
Error: Package: zfs-initramfs-2.0.0-rc1_30_gfd20a81.el7.x86_64 (/zfs-initramfs-2.0.0-rc1_30_gfd20a81.el7.x86_64)
           Requires: initramfs-tools
```
```
[user1@s1-test zfs]$ git status
# On branch zfs-2.0-release
nothing to commit, working directory clean
```
```
[user1@s1-test zfs]$ git log --oneline
fd20a81 Links in Source Files
```
```
[user1@s1-test zfs]$ ls
aclocal.m4          COPYRIGHT                                           Makefile.in                                         zfs-2.0.0-rc1_30_gfd20a81.el7.src.rpm
AUTHORS             cppcheck-suppressions.txt                           man                                                 zfs-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm
autogen.sh          etc                                                 META                                                zfs-2.0.0.tar.gz
bin                 include                                             module                                              zfs_config.h
build               lib                                                 NEWS                                                zfs_config.h.in
cmd                 libnvpair1-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm     NOTICE                                              zfs_config.h.in~
CODE_OF_CONDUCT.md  libtool                                             python2-pyzfs-2.0.0-rc1_30_gfd20a81.el7.noarch.rpm  zfs-debuginfo-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm
config              libuutil1-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm      README.md                                           zfs-dkms-2.0.0-rc1_30_gfd20a81.el7.noarch.rpm
config.log          libzfs2-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm        rpm                                                 zfs-dkms-2.0.0-rc1_30_gfd20a81.el7.src.rpm
config.status       libzfs2-devel-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm  scripts                                             zfs-dracut-2.0.0-rc1_30_gfd20a81.el7.noarch.rpm
configure           libzpool2-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm      stamp-h1                                            zfs-initramfs-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm
configure.ac        LICENSE                                             TEST                                                zfs.release
contrib             Makefile                                            tests                                               zfs.release.in
copy-builtin        Makefile.am                                         udev                                                zfs-test-2.0.0-rc1_30_gfd20a81.el7.x86_64.rpm
```

**Comment by tonymmm1** [comment]
Well I will wait and work on this later on maybe I will look through the makefiles and see what is happening.

**Comment by Ornias1993** [comment]
The can be the only post 0.8.4 that could be related...
https://github.com/openzfs/zfs/pull/10518

**Comment by tonymmm1** [comment]
Yeah that looks like it could be an issue in the configure I guess.

**Comment by Ornias1993** [comment]
It was also not tested with RPM's it seems:

> I have not tested that this definition of RPM_DEFINE_UTIL works, as I do not use RPM. It is possible that the situation has changed since this bit was originally written, and these defines must now not be provided for whatever reason. This should be tested by someone who does use RPM.

**Comment by tonymmm1** [comment]
Well thanks for your feedback I am signing off for now.

**Comment by Ornias1993** [comment]
@behlendorf I can confirm this is not just a documentation issue. 
Behavior did change in the past months and there is nothing to indicate a change in CentOS that would invalidate the documentation or create any error.
Suggesting "Type: Building" and "Type: Defect"

**Comment by hvdijk** [comment]
> The can be the only post 0.8.4 that could be related...
> #10518
> 
> Specifically:
> `RPM_DEFINE_UTIL=${RPM_DEFINE_UTIL}' $(DEFINE_INITRAMFS)'`

It's possible the problem only shows up after my commit, but I do not believe this to be the problem. My commit changed
```
       RPM_DEFINE_UTIL+=' $(DEFINE_INITRAMFS)'
```
to
```
       RPM_DEFINE_UTIL=${RPM_DEFINE_UTIL}' $(DEFINE_INITRAMFS)'
```
With bash (used by CentOS), these two lines mean the same thing unless there is a typo in there that I am not seeing even now. This part of the commit was just a portability fix that should have no impact on CentOS.

I'm not saying my commit is not responsible for this though. There were a small number of changes in my commit (included in the PR description) that were not just portability fixes. It is possible that those changes caused this, in which case my apologies for the breakage.

**Comment by Ornias1993** [comment]
@hvdijk Niet jouw fout (als dit al een fout is), had er dan in review uitgepakt moeten worden ;-)
TLDR: Don't worry about it

**Comment by hvdijk** [comment]
I checked now before and after my commit in a CentOS chroot. Both before and after, I see the zfs-initramfs RPM built only if `/usr/share/initramfs-tools` exists, and both before and after, it will fail to install if it is built. Perhaps there is some change in some other CentOS package that now causes `/usr/share/initramfs-tools` to be created on some systems?

As far as I can find, no rpm-based distro uses initramfs-tools. Perhaps the support for that should just be removed (and, of course, kept for deb-based distros)? That would solve this problem.

**Comment by Ornias1993** [comment]
@hvdijk I agree, im actually amazed by the fact its flagged as a dependency on nom-deb systems (hence rpm builds) to begin with...

**Comment by hvdijk** [comment]
I have created PR #10898 for that.
